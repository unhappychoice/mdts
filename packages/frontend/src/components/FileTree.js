"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const icons_material_1 = require("@mui/icons-material");
const ChevronLeft_1 = __importDefault(require("@mui/icons-material/ChevronLeft"));
const ChevronRight_1 = __importDefault(require("@mui/icons-material/ChevronRight"));
const Description_1 = __importDefault(require("@mui/icons-material/Description"));
const ExpandMore_1 = __importDefault(require("@mui/icons-material/ExpandMore"));
const UnfoldLess_1 = __importDefault(require("@mui/icons-material/UnfoldLess"));
const UnfoldMore_1 = __importDefault(require("@mui/icons-material/UnfoldMore"));
const material_1 = require("@mui/material");
const x_tree_view_1 = require("@mui/x-tree-view");
const SimpleTreeView_1 = require("@mui/x-tree-view/SimpleTreeView");
const react_1 = __importDefault(require("react"));
const useFileTree_1 = require("../hooks/apis/useFileTree");
const renderTreeItems = (tree, onFileSelect, parentPath = '') => {
    return tree.map((item) => {
        if (typeof item === 'string') {
            const fileName = item.split('/').pop();
            return ((0, jsx_runtime_1.jsx)(x_tree_view_1.TreeItem, { itemId: item, label: (0, jsx_runtime_1.jsx)(material_1.Typography, { variant: "body2", sx: { fontSize: '0.875rem' }, children: fileName }), icon: (0, jsx_runtime_1.jsx)(Description_1.default, { fontSize: "small" }), onClick: () => onFileSelect(item) }, item));
        }
        else {
            const key = Object.keys(item)[0];
            const value = item[key];
            const currentPath = parentPath ? `${parentPath}/${key}` : key;
            return ((0, jsx_runtime_1.jsx)(x_tree_view_1.TreeItem, { itemId: currentPath, label: (0, jsx_runtime_1.jsxs)(material_1.Box, { sx: { display: 'flex', alignItems: 'center' }, children: [(0, jsx_runtime_1.jsx)(icons_material_1.FolderOutlined, { sx: { mr: 1, fontSize: 'small' }, color: "primary" }), (0, jsx_runtime_1.jsx)(material_1.Typography, { variant: "body2", sx: { fontSize: '0.875rem' }, children: key })] }), children: Array.isArray(value) && renderTreeItems(value, onFileSelect, currentPath) }, currentPath));
        }
    });
};
const FileTree = ({ onFileSelect, isOpen, onToggle, selectedFilePath }) => {
    const { fileTree, loading, error } = (0, useFileTree_1.useFileTree)();
    const [expanded, setExpanded] = react_1.default.useState([]);
    const handleExpandAll = () => {
        const allItemIds = [];
        const collectIds = (items, parentPath = '') => {
            items.forEach(item => {
                if (typeof item !== 'string') {
                    const key = Object.keys(item)[0];
                    const currentPath = parentPath ? `${parentPath}/${key}` : key;
                    allItemIds.push(currentPath);
                    collectIds(item[key], currentPath);
                }
            });
        };
        if (fileTree) {
            collectIds(fileTree);
        }
        setExpanded(allItemIds);
    };
    const handleCollapseAll = () => {
        setExpanded([]);
    };
    return ((0, jsx_runtime_1.jsxs)(material_1.Box, { sx: {
            width: isOpen ? '300px' : '66px',
            bgcolor: 'background.paper',
            p: isOpen ? 2 : 0.5,
            borderRight: '1px solid',
            borderColor: 'divider',
            minHeight: '100%',
            flexShrink: 0,
        }, children: [(0, jsx_runtime_1.jsxs)(material_1.Box, { sx: { display: 'flex', justifyContent: 'space-between', alignItems: 'center', height: '34px', marginTop: isOpen ? '0' : '16px', marginBottom: 2 }, children: [isOpen && ((0, jsx_runtime_1.jsxs)(material_1.Box, { sx: { display: 'flex', alignItems: 'center', flex: 1 }, children: [(0, jsx_runtime_1.jsx)(material_1.Typography, { variant: "h6", gutterBottom: true, sx: { flex: 1, marginLeft: 1, marginBottom: 0 }, children: "File Tree" }), (0, jsx_runtime_1.jsx)(material_1.IconButton, { onClick: handleExpandAll, size: "small", "aria-label": "expand all", children: (0, jsx_runtime_1.jsx)(UnfoldMore_1.default, {}) }), (0, jsx_runtime_1.jsx)(material_1.IconButton, { onClick: handleCollapseAll, size: "small", "aria-label": "collapse all", children: (0, jsx_runtime_1.jsx)(UnfoldLess_1.default, {}) })] })), (0, jsx_runtime_1.jsx)(material_1.IconButton, { onClick: onToggle, size: "small", sx: { marginBottom: 0, marginLeft: isOpen ? '0' : '12px' }, children: isOpen ? (0, jsx_runtime_1.jsx)(ChevronLeft_1.default, {}) : (0, jsx_runtime_1.jsx)(ChevronRight_1.default, {}) })] }), isOpen && (loading ? ((0, jsx_runtime_1.jsx)(material_1.Box, { sx: { display: 'flex', justifyContent: 'center', alignItems: 'center', height: '100vh' }, children: (0, jsx_runtime_1.jsx)(material_1.CircularProgress, {}) })) : error ? ((0, jsx_runtime_1.jsxs)(material_1.Typography, { color: "error", children: ["Error: ", error] })) : ((0, jsx_runtime_1.jsx)(SimpleTreeView_1.SimpleTreeView, { defaultCollapseIcon: (0, jsx_runtime_1.jsx)(ExpandMore_1.default, {}), defaultExpandIcon: (0, jsx_runtime_1.jsx)(ChevronRight_1.default, {}), expandedItems: expanded, onExpandedItemsChange: (event, itemIds) => setExpanded(itemIds), sx: { flexGrow: 1, maxWidth: 400, overflowY: 'auto' }, children: renderTreeItems(fileTree, onFileSelect) })))] }));
};
exports.default = FileTree;
