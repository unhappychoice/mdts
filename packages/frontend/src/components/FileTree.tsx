import { ArticleOutlined, Clear as ClearIcon, FolderOutlined } from '@mui/icons-material';
import ChevronLeftIcon from '@mui/icons-material/ChevronLeft';
import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import DescriptionIcon from '@mui/icons-material/Description';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import UnfoldLessIcon from '@mui/icons-material/UnfoldLess';
import UnfoldMoreIcon from '@mui/icons-material/UnfoldMore';
import { Box, CircularProgress, IconButton, InputBase, InputAdornment, Typography } from '@mui/material';
import { TreeItem } from '@mui/x-tree-view';
import { SimpleTreeView } from '@mui/x-tree-view/SimpleTreeView';
import React, { useEffect, useMemo } from 'react';
import { useFileTree } from '../hooks/apis/useFileTree';

interface FileTreeItem {
  [key: string]: (FileTreeItem | string)[];
}

interface FileTreeComponentProps {
  onFileSelect: (path: string) => void;
  isOpen: boolean;
  onToggle: () => void;
  selectedFilePath: string | null;
}

const FileTree: React.FC<FileTreeComponentProps> = ({ onFileSelect, isOpen, onToggle }) => {
  const { fileTree, loading, error } = useFileTree();
  const [expanded, setExpanded] = React.useState<string[]>([]);
  const [searchQuery, setSearchQuery] = React.useState<string>('');

  const filteredTree = useMemo(() => filterTree(fileTree, searchQuery), [fileTree, searchQuery]);

  const handleSearchChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(event.target.value);
  };

  const handleExpandAll = () => {
    const allItemIds: string[] = [];
    const collectIds = (items: (FileTreeItem | string)[], parentPath: string = '') => {
      items.forEach(item => {
        if (typeof item !== 'string') {
          const key = Object.keys(item)[0];
          const currentPath = parentPath ? `${parentPath}/${key}` : key;
          allItemIds.push(currentPath);
          collectIds(item[key] as FileTreeItem[] | string[], currentPath);
        }
      });
    };
    if (fileTree) {
      collectIds(fileTree);
    }
    setExpanded(allItemIds);
  };

  const handleCollapseAll = () => {
    setExpanded([]);
  };

  useEffect(() => {
    if (searchQuery) {
      const newExpanded: string[] = [];

      const collectExpandedPaths = (items: (FileTreeItem | string)[], parentPath: string = '') => {
        items.forEach(item => {
          if (typeof item !== 'string') { // It's a folder
            const key = Object.keys(item)[0];
            const value = item[key];
            const currentPath = parentPath ? `${parentPath}/${key}` : key;
            if (!newExpanded.includes(currentPath)) {
              newExpanded.push(currentPath);
            }
            if (Array.isArray(value)) {
              collectExpandedPaths(value, currentPath);
            }
          }
        });
      };

      if (filteredTree) {
        collectExpandedPaths(filteredTree);
      }
      
      setExpanded(newExpanded);
    } else {
      setExpanded([]);
    }
  }, [searchQuery, filteredTree]);

  return (
    <Box sx={{
      width: isOpen ? '300px' : '66px',
      bgcolor: 'background.paper',
      p: isOpen ? 2 : 0.5,
      borderRight: '1px solid',
      borderColor: 'divider',
      minHeight: '100%',
      flexShrink: 0,
    }}>
      <Box sx={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', height: '34px', marginTop: isOpen ? '0' : '16px', marginBottom: 2 }}>
        {isOpen && (
          <Box sx={{ display: 'flex', alignItems: 'center', flex: 1 }}>
            <Typography variant="h6" gutterBottom sx={{ flex: 1, marginLeft: 1, marginBottom: 0 }}>
              File Tree
            </Typography>
            <IconButton onClick={handleExpandAll} size="small" aria-label="expand all">
              <UnfoldMoreIcon />
            </IconButton>
            <IconButton onClick={handleCollapseAll} size="small" aria-label="collapse all">
              <UnfoldLessIcon />
            </IconButton>
          </Box>
        )}
        <IconButton onClick={onToggle} size="small" sx={{ marginBottom: 0, marginLeft: isOpen ? '0' : '12px' }}>
          {isOpen ? <ChevronLeftIcon /> : <ChevronRightIcon />}
        </IconButton>
      </Box>
      <Box mb={2}>
        <InputBase
          placeholder="Search files..."
          inputProps={{ 'aria-label': 'search files' }}
          sx={{
            flex: 1,
            width: '100%',
            px: 2,
            py: 0,
            border: '1px solid',
            borderColor: 'divider',
            borderRadius: '4px',
            fontSize: '0.875rem',
          }}
          value={searchQuery}
          onChange={handleSearchChange}
          endAdornment={
            searchQuery && (
              <InputAdornment position="end">
                <IconButton
                  onClick={() => setSearchQuery('')}
                  edge="end"
                  size="small"
                  aria-label="clear search"
                >
                  <ClearIcon sx={{ fontSize: '0.875rem', color: 'divider' }}/>
                </IconButton>
              </InputAdornment>
            )
          }
        />
      </Box>
      {isOpen && (
          loading ? (
            <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '100vh' }}>
              <CircularProgress />
            </Box>
          ) : error ? (
            <Typography color="error">Error: {error}</Typography>
          ) : (
            <SimpleTreeView
              defaultCollapseIcon={<ExpandMoreIcon />}
              defaultExpandIcon={<ChevronRightIcon />}
              expandedItems={expanded}
              onExpandedItemsChange={(event, itemIds) => setExpanded(itemIds)}
              sx={{ flexGrow: 1, maxWidth: 400, overflowY: 'auto' }}
            >
              {renderTreeItems(filteredTree, onFileSelect)}
            </SimpleTreeView>
          )
      )}
    </Box>
  );
};

export default FileTree;

const renderTreeItems = (
  tree: (FileTreeItem | string)[],
  onFileSelect: (path: string) => void,
  parentPath: string = ''
) => {
  return tree.map((item) => {
    if (typeof item === 'string') {
      const fileName = item.split('/').pop();
      return (
        <TreeItem
          key={item}
          itemId={item}
          label={
            <Box sx={{ display: 'flex', alignItems: 'center' }}>
              <ArticleOutlined sx={{ mr: 1, fontSize: 'small' }} />
              <Typography variant="body2" sx={{ fontSize: '0.875rem' }}>
                {fileName}
              </Typography>
            </Box>
          }
          onClick={() => onFileSelect(item)}
        />
      );
    } else {
      const key = Object.keys(item)[0];
      const value = item[key];
      const currentPath = parentPath ? `${parentPath}/${key}` : key;
      return (
        <TreeItem
          key={currentPath}
          itemId={currentPath}
          label={
            <Box sx={{ display: 'flex', alignItems: 'center' }}>
              <FolderOutlined sx={{ mr: 1, fontSize: 'small' }} color="primary" />
              <Typography variant="body2" sx={{ fontSize: '0.875rem' }}>
                {key}
              </Typography>
            </Box>
          }
        >
          {Array.isArray(value) && value.length > 0 && renderTreeItems(value, onFileSelect, currentPath)}
        </TreeItem>
      );
    }
  });
};

const filterTree = (tree: (FileTreeItem | string)[], searchQuery: string): (FileTreeItem | string)[] => {
  if (!searchQuery) return tree; // If no search query, return the original tree

  return tree
    .reduce((acc: (FileTreeItem | string)[], item) => {
      if (typeof item === 'string') {
        const fileName = item.split('/').pop() || '';
        return fileName.toLowerCase().includes(searchQuery.toLowerCase())
          ? [...acc, item]
          : acc;
      } else {
        const key = Object.keys(item)[0];
        const value = item[key];

        const children = Array.isArray(value)
          ? filterTree(value, searchQuery)
          : [];

        return children.length > 0
          ? [...acc, { [key]: children } as FileTreeItem ]
          :acc;
      }
    }, []);
};
