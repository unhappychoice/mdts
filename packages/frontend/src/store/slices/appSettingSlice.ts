import { createSlice } from '@reduxjs/toolkit';

type DarkMode = 'dark' | 'light' | 'auto';
type ContentMode = 'full' | 'compact';

interface AppSettingState {
  darkMode: DarkMode;
  contentMode: ContentMode;
  fileTreeOpen: boolean;
  outlineOpen: boolean;
}

const initialState: AppSettingState = {
  darkMode: 'auto',
  contentMode: 'compact',
  fileTreeOpen: true,
  outlineOpen: true,
};

const appSettingSlice = createSlice({
  name: 'appSetting',
  initialState,
  reducers: {
    saveAppSetting: (state, action: PayloadAction<{ darkMode: DarkMode, contentMode: ContentMode }>) => {
      state.darkMode = action.payload.darkMode;
      state.contentMode = action.payload.contentMode;
    },
    toggleFileTree: (state) => {
      state.fileTreeOpen = !state.fileTreeOpen;
    },
    toggleOutline: (state) => {
      state.outlineOpen = !state.outlineOpen;
    },
  },
});

export const { saveAppSetting, toggleFileTree, toggleOutline } = appSettingSlice.actions;

export default appSettingSlice.reducer;
