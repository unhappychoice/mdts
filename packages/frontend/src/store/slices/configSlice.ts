import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

export const fetchConfig = createAsyncThunk('config/fetchConfig', async () => {
  return await fetchData<ConfigState>('/api/config', 'json');
});

export const saveConfigToBackend = createAsyncThunk(
  'config/saveConfigToBackend',
  async (config: ConfigState, { dispatch }) => {
    try {
      await fetch('/api/config', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(config),
      });
      dispatch(fetchConfig()); // Re-fetch config after saving
    } catch (error) {
      console.error('Failed to save settings:', error);
      throw error; // Re-throw to allow error handling in component
    }
  }
);

export interface ConfigState {
  fontFamily: string;
  fontFamilyMonospace: string;
  fontSize: number;
  syntaxHighlighterTheme: string;
}

const initialState: ConfigState = {
  fontFamily: 'Roboto',
  fontFamilyMonospace: 'monospace',
  fontSize: 14,
  syntaxHighlighterTheme: 'auto',
};

const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchConfig.fulfilled, (state, action: PayloadAction<ConfigState>) => {
      state.fontFamily = action.payload.fontFamily;
      state.fontFamilyMonospace = action.payload.fontFamilyMonospace;
      state.fontSize = action.payload.fontSize;
      state.syntaxHighlighterTheme = action.payload.syntaxHighlighterTheme;
    });
  },
});

export default configSlice.reducer;
