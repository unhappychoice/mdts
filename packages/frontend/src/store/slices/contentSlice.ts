
import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

interface ContentState {
  content: string;
  loading: boolean;
  error: string | null;
  scrollPosition: number;
}

const initialState: ContentState = {
  content: '',
  loading: true,
  error: null,
  scrollPosition: 0,
};

export const fetchContent = createAsyncThunk(
  'content/fetchContent',
  async (path: string | null) => {
    const url = path ? `/api/markdown/${path}` : '/api/markdown/mdts-welcome-markdown.md';
    const data = await fetchData<string>(url, 'text');
    return data || '';
  }
);

const contentSlice = createSlice({
  name: 'content',
  initialState,
  reducers: {
    setScrollPosition: (state, action: PayloadAction<number>) => {
      state.scrollPosition = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchContent.pending, (state) => {
        if (!state.content) {
          state.loading = true;
        }
        state.error = null;
      })
      .addCase(fetchContent.fulfilled, (state, action) => {
        state.loading = false;
        state.content = action.payload;
      })
      .addCase(fetchContent.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message || 'Failed to fetch content';
      });
  },
});

export const { setScrollPosition } = contentSlice.actions;

export default contentSlice.reducer;
