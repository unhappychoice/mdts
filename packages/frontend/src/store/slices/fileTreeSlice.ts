
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

interface FileTreeItem {
  [key: string]: (FileTreeItem | string)[];
}

interface FileTreeState {
  fileTree: (FileTreeItem | string)[];
  filteredFileTree: (FileTreeItem | string)[];
  searchQuery: string;
  expandedNodes: string[];
  mountedDirectoryPath: string;
  loading: boolean;
  error: string | null;
}

const initialState: FileTreeState = {
  fileTree: [],
  filteredFileTree: [],
  searchQuery: '',
  expandedNodes: [],
  mountedDirectoryPath: '',
  loading: true,
  error: null,
};

const LOCAL_STORAGE_KEY_PREFIX = 'mdts_expanded_nodes_';
const LOCAL_STORAGE_RECENT_PATHS_KEY = 'mdts_recent_paths';
const MAX_RECENT_PATHS = 10;

const saveExpandedNodes = (path: string, nodes: string[]) => {
  try {
    localStorage.setItem(`${LOCAL_STORAGE_KEY_PREFIX}${path}`, JSON.stringify(nodes));
    let recentPaths: string[] = JSON.parse(localStorage.getItem(LOCAL_STORAGE_RECENT_PATHS_KEY) || '[]');
    recentPaths = recentPaths.filter(p => p !== path);
    recentPaths.unshift(path);
    if (recentPaths.length > MAX_RECENT_PATHS) {
      const oldPath = recentPaths.pop();
      if (oldPath) {
        localStorage.removeItem(`${LOCAL_STORAGE_KEY_PREFIX}${oldPath}`);
      }
    }
    localStorage.setItem(LOCAL_STORAGE_RECENT_PATHS_KEY, JSON.stringify(recentPaths));
  } catch (e) {
    console.error('Failed to save expanded nodes to local storage', e);
  }
};

const loadExpandedNodes = (path: string): string[] => {
  try {
    const storedNodes = localStorage.getItem(`${LOCAL_STORAGE_KEY_PREFIX}${path}`);
    return storedNodes ? JSON.parse(storedNodes) : [];
  } catch (e) {
    console.error('Failed to load expanded nodes from local storage', e);
    return [];
  }
};

const filterTree = (tree: (FileTreeItem | string)[], searchQuery: string): (FileTreeItem | string)[] => {
  if (!searchQuery) return tree;

  return tree
    .reduce((acc: (FileTreeItem | string)[], item) => {
      if (typeof item === 'string') {
        const fileName = item.split('/').pop() || '';
        return fileName.toLowerCase().includes(searchQuery.toLowerCase())
          ? [...acc, item]
          : acc;
      } else {
        const key = Object.keys(item)[0];
        const value = item[key];

        const children = Array.isArray(value)
          ? filterTree(value, searchQuery)
          : [];

        return children.length > 0
          ? [...acc, { [key]: children } as FileTreeItem ]
          :acc;
      }
    }, []);
};

export const fetchFileTree = createAsyncThunk(
  'fileTree/fetchFileTree',
  async () => {
    const data = await fetchData<{ fileTree: FileTreeItem[], mountedDirectoryPath: string }>('/api/filetree', 'json');
    return { fileTree: data?.fileTree || [], mountedDirectoryPath: data?.mountedDirectoryPath };
  }
);

const fileTreeSlice = createSlice({
  name: 'fileTree',
  initialState,
  reducers: {
    setSearchQuery: (state, action: { payload: string }) => {
      state.searchQuery = action.payload;
      state.filteredFileTree = filterTree(state.fileTree, action.payload);
    },
    toggleNode: (state, action: { payload: string }) => {
      const path = action.payload;
      if (state.expandedNodes.includes(path)) {
        state.expandedNodes = state.expandedNodes.filter((nodePath) => nodePath !== path);
      } else {
        state.expandedNodes.push(path);
      }
      saveExpandedNodes(state.mountedDirectoryPath, state.expandedNodes);
    },
    setExpandedNodes: (state, action: { payload: string[] }) => {
      state.expandedNodes = action.payload;
      saveExpandedNodes(state.mountedDirectoryPath, state.expandedNodes);
    },
    expandAllNodes: (state, action: { payload: (FileTreeItem | string)[] | null }) => {
      const allItemIds: string[] = [];
      const collectIds = (items: (FileTreeItem | string)[], parentPath: string = '') => {
        items.forEach(item => {
          if (typeof item !== 'string') {
            const key = Object.keys(item)[0];
            const currentPath = parentPath ? `${parentPath}/${key}` : key;
            allItemIds.push(currentPath);
            collectIds(item[key] as (FileTreeItem | string)[], currentPath);
          }
        });
      };
      if (action.payload) {
        collectIds(action.payload);
      }
      state.expandedNodes = allItemIds;
      saveExpandedNodes(state.mountedDirectoryPath, state.expandedNodes);
    },
    setMountedDirectoryPath: (state, action: { payload: string }) => {
      state.mountedDirectoryPath = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchFileTree.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchFileTree.fulfilled, (state, action) => {
        state.loading = false;
        state.fileTree = action.payload.fileTree;
        state.mountedDirectoryPath = action.payload.mountedDirectoryPath;
        state.filteredFileTree = filterTree(action.payload.fileTree, state.searchQuery);
        state.expandedNodes = loadExpandedNodes(action.payload.mountedDirectoryPath);
      })
      .addCase(fetchFileTree.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message || 'Failed to fetch file tree';
      });
  },
});

export const selectFilteredFileTree = (
  fullFileTree: (FileTreeItem | string)[],
  targetPath: string
): (FileTreeItem | string)[] => {
  if (targetPath === '') {
    const result: FileTreeItem[] | string[] = [];
    fullFileTree.forEach(item => {
      if (typeof item === 'string') {
        result.push(item.split('/').pop() || item);
      } else {
        const key = Object.keys(item)[0];
        const newObject: FileTreeItem = {};
        newObject[key.split('/').pop() || key] = item[key];
        result.push(newObject);
      }
    });
    return result;
  }

  const findChildren = (
    currentTree: (FileTreeItem | string)[],
    pathSegments: string[],
    currentSegmentIndex: number
  ): (FileTreeItem | string)[] | null => {
    if (currentSegmentIndex === pathSegments.length) {
      return currentTree;
    }

    const segment = pathSegments[currentSegmentIndex];

    for (const item of currentTree) {
      if (typeof item !== 'string') {
        const key = Object.keys(item)[0];
        const itemSegments = key.split('/');
        if (itemSegments[itemSegments.length - 1] === segment) {
          const children = item[key];
          if (Array.isArray(children)) {
            return findChildren(children, pathSegments, currentSegmentIndex + 1);
          }
        }
      }
    }
    return null;
  };

  const pathSegments = targetPath.split('/').filter(s => s !== '');
  const children = findChildren(fullFileTree, pathSegments, 0);

  if (children) {
    const result: FileTreeItem[] | string[] = [];
    children.forEach(item => {
      if (typeof item === 'string') {
        result.push(item.split('/').pop() || item);
      } else {
        const key = Object.keys(item)[0];
        const newObject: FileTreeItem = {};
        newObject[key.split('/').pop() || key] = item[key];
        result.push(newObject);
      }
    });
    return result;
  }

  return [];
};

export const {
  setSearchQuery,
  toggleNode,
  setExpandedNodes,
  expandAllNodes,
  setMountedDirectoryPath,
} = fileTreeSlice.actions;

export default fileTreeSlice.reducer;
