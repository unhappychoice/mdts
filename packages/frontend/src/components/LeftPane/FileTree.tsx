import { Box } from '@mui/material';
import React, { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchFileTree, setExpandedNodes, setSearchQuery, expandAllNodes, FileTreeItem } from '../../store/slices/fileTreeSlice';
import { AppDispatch, RootState } from '../../store/store';
import FileTreeHeader from './FileTreeHeader';
import FileTreeSearch from './FileTreeSearch';
import FileTreeContent from './FileTreeContent';

interface FileTreeComponentProps {
  onFileSelect: (path: string) => void;
  isOpen: boolean;
  onToggle: () => void;
  selectedFilePath: string | null;
}

const FileTree: React.FC<FileTreeComponentProps> = ({ onFileSelect, isOpen, onToggle }) => {
  const dispatch = useDispatch<AppDispatch>();
  const { fileTree, filteredFileTree, loading, error, searchQuery, expandedNodes } = useSelector((state: RootState) => state.fileTree);

  useEffect(() => {
    dispatch(fetchFileTree());
  }, [dispatch]);

  const handleSearchChange = useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(setSearchQuery(event.target.value));
  }, [dispatch]);

  const handleClearSearch = useCallback(() => {
    dispatch(setSearchQuery(''));
  }, [dispatch]);

  const handleExpandAllClick = useCallback(() => {
    dispatch(expandAllNodes(fileTree));
  }, [dispatch, fileTree]);

  const handleCollapseAll = useCallback(() => {
    dispatch(setExpandedNodes([]));
  }, [dispatch]);

  const handleExpandedItemsChange = useCallback((event: React.SyntheticEvent, itemIds: string[]) => {
    dispatch(setExpandedNodes(itemIds));
  }, [dispatch]);

  useEffect(() => {
    if (!searchQuery) return;

    const newExpanded: string[] = [];

    const collectExpandedPaths = (items: (FileTreeItem | { [key: string]: (FileTreeItem | object)[] })[], parentPath: string = '') => {
      items.forEach(item => {
        if (!('path' in item)) { // It's a folder
          const key = Object.keys(item)[0];
          const value = (item as { [key: string]: (FileTreeItem | object)[] })[key];
          const currentPath = parentPath ? `${parentPath}/${key}` : key;
          if (!newExpanded.includes(currentPath)) {
            newExpanded.push(currentPath);
          }
          if (Array.isArray(value)) {
            collectExpandedPaths(value, currentPath);
          }
        }
      });
    };

    if (filteredFileTree) {
      collectExpandedPaths(filteredFileTree);
    }

    dispatch(setExpandedNodes(newExpanded));
  }, [searchQuery, filteredFileTree, dispatch]);

  return (
    <Box sx={{
      width: isOpen ? '300px' : '66px',
      bgcolor: 'background.paper',
      py: 2,
      borderRight: '1px solid',
      borderColor: 'divider',
      minHeight: '100%',
      flexShrink: 0,
    }}>
      <FileTreeHeader
        isOpen={isOpen}
        onToggle={onToggle}
        onExpandAllClick={handleExpandAllClick}
        onCollapseAll={handleCollapseAll}
      />
      {isOpen && (
        <FileTreeSearch
          searchQuery={searchQuery}
          onSearchChange={handleSearchChange}
          onClearSearch={handleClearSearch}
        />
      )}
      {isOpen && (
        <FileTreeContent
          filteredFileTree={filteredFileTree}
          loading={loading}
          error={error}
          expandedNodes={expandedNodes}
          onFileSelect={onFileSelect}
          onExpandedItemsChange={handleExpandedItemsChange}
          dispatch={dispatch}
        />
      )}
    </Box>
  );
};

export default FileTree;

