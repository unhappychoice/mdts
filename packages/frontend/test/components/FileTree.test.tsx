import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import FileTree from '../../src/components/FileTree';
import { fetchFileTree, setSearchQuery } from '../../src/store/slices/fileTreeSlice';

const mockStore = configureStore([]);

describe('FileTree', () => {
  let store;
  const initialState = {
    fileTree: {
      fileTree: [
        { 'folder1': ['/folder1/file1.md', '/folder1/file2.txt'] },
        '/file3.js',
      ],
      filteredFileTree: [
        { 'folder1': ['/folder1/file1.md', '/folder1/file2.txt'] },
        '/file3.js',
      ],
      loading: false,
      error: null,
      searchQuery: '',
    },
  };

  beforeEach(() => {
    store = mockStore(initialState);
    store.dispatch = jest.fn();
  });

  test('renders correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('dispatches fetchFileTree on mount', () => {
    render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );
    expect(store.dispatch).toHaveBeenCalledWith(fetchFileTree());
  });

  test('dispatches setSearchQuery on search input change', () => {
    render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );

    const searchInput = screen.getByPlaceholderText('Search files...');
    fireEvent.change(searchInput, { target: { value: 'file' } });

    expect(store.dispatch).toHaveBeenCalledWith(setSearchQuery('file'));
  });

  test('clears search query when clear button is clicked', () => {
    store = mockStore({
      fileTree: {
        ...initialState.fileTree,
        searchQuery: 'file',
        filteredFileTree: ['/folder1/file1.md'],
      },
    });
    store.dispatch = jest.fn();

    render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );

    const clearButton = screen.getByLabelText('clear search');
    fireEvent.click(clearButton);

    expect(store.dispatch).toHaveBeenCalledWith(setSearchQuery(''));
  });

  test('expands all folders when expand all button is clicked', () => {
    render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );

    const expandAllButton = screen.getByLabelText('expand all');
    fireEvent.click(expandAllButton);

    // This test needs to assert on the expanded state of the TreeView, which is internal to the component.
    // For now, we'll just ensure the button click doesn't cause an error.
    // A more robust test would involve checking the `expandedItems` prop of SimpleTreeView.
  });

  test('collapses all folders when collapse all button is clicked', () => {
    render(
      <Provider store={store}>
        <FileTree onFileSelect={jest.fn()} isOpen={true} onToggle={jest.fn()} selectedFilePath={null} />
      </Provider>
    );

    const collapseAllButton = screen.getByLabelText('collapse all');
    fireEvent.click(collapseAllButton);

    // Similar to expand all, this would require inspecting internal state.
  });
});
