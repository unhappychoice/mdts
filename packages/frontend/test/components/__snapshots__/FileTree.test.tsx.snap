// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FileTree renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1kgtqjm"
  >
    <div
      class="MuiBox-root css-swwgw5"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-130mf42-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ChevronRightIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
